Fixed the settings issue

wokhan opened the pull request:
Now also offering a portable mode; using .Net Core embedded conf management feature + WPF designer (best of both worlds).

Options descriptions will be improved in a later change (along with UI improvements).

Note: config tests have been removed (checking for settings file availability isn't mandatory anymore as default values will be taken directly from the assembly, and default values testing wasn't testing default values, but data saving without mocking the output file). I'll rewrite then someday...

harrwiss commented:
Hi @wokhan, thx for taking care of this :)

However, when I try to run the Console in VS it doesn't start:

NRE in app.xml.cs:
` if (Settings.Default.AlwaysRunAs && !UAC.CheckProcessElevated())`

Settings is null at this point, because ins Settings.cs the following line returns null:
`defaultInstance = configuration.Get<Settings>();`

Did you test this?

wokhan commented:
Hi @harrwiss,
Of course I did but... it was before I added this portable mode option, sorry about that. It appears .Net Core settings management can send back null instead of default values when no file is provided. Since at first I had one, it got saved and was "preventing" this issue afterwards.
Reproduced and fixed! (now the default values are kept if config.Get doesn't return null).
Thanks

wokhan commented:
NB : now only user scoped attributes are actually saved to the settings file. There is no point in saving application settings (which, as you mentioned, are not supposed to change). I don't know why MS implemented this that way in WPF...

wokhan commented:
```
Settings newSettings = configuration.Get<Settings>();
if (newSettings != null)
{
      defaultInstance = newSettings;
}
```

is the exact same thing than:

`defaultInstance = configuration.Get<Settings>() ?? defaultInstance;`

Did you try with the second commit I made right after the first ? 🙄 
First fix was totally wrong.

harrwiss commented:
No, I did not see that there is another one 😏 - looks ok now
Will do some further test....

wokhan commented:
Thank you and sorry for this double mistake 😄

harrwiss commented:
Sorry, think I closed the pull request by mistake - re-opening again....

harrwiss commented on Console/UI/Pages/Options.xaml.cs:
```diff
@@ -72,19 +59,9 @@ private void btnResetDefault_Click(object sender, RoutedEventArgs e)
-        private void txtSharedConfigurationPath_MouseDown(object sender, System.Windows.Input.MouseButtonEventArgs e)
-        {
-            ProcessHelper.StartShellExecutable("explorer.exe", SharedConfigurationPath, true);
+            ProcessHelper.StartShellExecutable("explorer.exe", $"/select \"{Settings.Default.ConfigurationPath}\"", true);
```
The "/select" option doesn't open the file:
`ProcessHelper.StartShellExecutable("explorer.exe", $"/select \"{Settings.Default.ConfigurationPath}\"", true);`

This way it works correctly (as before):
ProcessHelper.StartShellExecutable("explorer.exe", $"\"{Settings.Default.ConfigurationPath}\"", true);

harrwiss replied:
I pushed this small fix

wokhan replied:
I didn't want the file to directly open, but to be located (hence the select, which you are write, isn't working - I don't know why as it's pretty standard).
Directly opening the file doesn't seem OK for me as I usually use multiple editors depending on my needs. Opening with the default might then be blocking.

harrwiss commented on Common/Config/Settings.Designer.cs:
```diff
@@ -1,10 +1,10 @@
-//     This code was generated by a tool.
-//     Runtime Version:4.0.30319.42000
+//     Ce code a été généré par un outil.
+//     Version du runtime :4.0.30319.42000
```
Would be nice to have comments always in english, but probably your windows is in french I suppose and this is aut-generated by VS.

wokhan replied:
Yes of course I never comment in French in my "personal" projects. I can switch my VS to english anyway, as translation isn't that good 😄

harrwiss commented on Common/Config/Settings.Designer.cs:
```diff
@@ -320,5 +313,17 @@ public sealed partial class Settings : global::System.Configuration.ApplicationS
+                this["IsPortable"] = value;
+            }
+        }
     }
```
Good to have 👍

wokhan replied:
Thanks 😄 
I guess we can also use a "log" subfolder in the application folder when portable mode is enabled (to ensure everything is indeed in the same place). This will be for a future branch...
